test/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
+        &name[..name.len() - 3]
+    }};
+}
+
+/// Creates a temporary file with the specified prefix in a directory named
+/// after the calling function, ie using it within a test will place it in a
+/// temporary directory named after the test
+#[macro_export]
+macro_rules! temp_file {
+    ($prefix:expr) => {{
+        let name = $crate::func_name!();
+        let name = name.strip_suffix("::{{closure}}").unwrap_or(name);
+        let mut name = name.replace("::", ".");
+        name.push('-');
+        name.push_str($prefix);
+        name.push('-');
+        tempfile::NamedTempFile::with_prefix(name).unwrap()
+    }};
+}
+
+pub use tracing::{subscriber::DefaultGuard, Level};
+
+pub fn init_logging(level: Level, test_pkg: &'static str) -> DefaultGuard {
+    use tracing_subscriber::{layer::SubscriberExt as _, Layer as _};
+    let layer = tracing_subscriber::fmt::layer()
+        .with_test_writer()
+        .with_filter(tracing_subscriber::filter::LevelFilter::from_level(level))
+        .with_filter(tracing_subscriber::EnvFilter::new(format!(
+            "{test_pkg}=trace,qt=trace,quilkin=trace"
+        )));
+    let sub = tracing_subscriber::Registry::default().with(layer);
+    let disp = tracing::dispatcher::Dispatch::new(sub);
+    let guard = tracing::dispatcher::set_default(&disp);
+
+    guard
+}
+
+#[macro_export]
+macro_rules! trace_test {
+    ($name:ident, $body:block) => {
+        #[tokio::test]
+        async fn $name() {
+            // Get the module name
+            let fname = $crate::func_name!();
+            let mname = fname.rsplit("::").nth(2).unwrap();
+
+            let _guard = init_logging($crate::Level::DEBUG, mname);
+
+            $body
+        }
+    };
+}
+
+pub struct ServerPailConfig {
+    pub packet_size: u16,
+    pub num_packets: Option<usize>,
+}
+
+impl Default for ServerPailConfig {
+    fn default() -> Self {
+        Self {
+            packet_size: 1024,
+            num_packets: None,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct AdminPailConfig;
+
+#[derive(Default)]
+pub struct RelayPailConfig {
+    pub config: quilkin::test::TestConfig,
+}
+
+#[derive(Default)]
+pub struct ProxyPailConfig {}
+
+#[derive(Default)]
+pub struct ManagementPailConfig {}
+
+#[derive(Default)]
+pub struct AgentPailConfig {
+    pub endpoints: Vec<(&'static str, &'static [&'static str])>,
+    pub icao_code: quilkin::config::IcaoCode,
+}
+
+pub enum PailConfig {
+    /// Creates a UDP socket with an ephemeral port that will send received
+    /// packets to a channel
+    Server(ServerPailConfig),
+    // Creates an admin server that is applied to any following relay, proxy, or management pails
+    //Admin(AdminPailConfig),
+    Relay(RelayPailConfig),
+    Proxy(ProxyPailConfig),
+    //Management(ManagementPailConfig),
```

Comment:
Are these commented out because they are unused?

Reply:
Yes, for now, I only did a few tests in this PR because it is already so big, and want to convert more in a separate PR.